### PR TITLE
epoll: handle edge case for epoll_ctl

### PR DESCRIPTION
**Pull request by tiif**
There is a test case that revealed that our implementation differs from the real system:
 - Set up an epoll watching the FD
- Call epoll_wait
- Set up another epoll watching the same FD
- Call epoll_wait on the first epoll. Nothing should be reported!

This happened because, in ``epoll_ctl``, we used ``check_and_update_readiness``, which is a function that would return notification for all epoll file description that registered a particular file description. But we shouldn't do that because no notification should be returned if there is no I/O activity between two ``epoll_wait`` (every first ``epoll_wait`` that happens after ``epoll_ctl`` is an exception, we should return notification that reflects the readiness of file description). 

r? @oli-obk

### CODE REVIEW

File: src/shims/unix/linux/epoll.rs [hunk 1]
```diff
@@ -525,3 +508,28 @@ fn ready_list_next(
...
+/// epoll_interest and, if necessary, triggers the notification. Unlike check_and_update_readiness,
+/// this function sends a notification to only one epoll instance.
+fn check_fd_ready_event(
+    fd_ref: &FileDescriptionRef,
+    interest: Rc<RefCell<EpollEventInterest>>,
+    id: FdId,
+    ecx: &MiriInterpCx<'_>,
+) {
+    // Get the bitmask of ready events for a file description.
+    let ready_events_bitmask = fd_ref.get_epoll_ready_events().unwrap().get_event_bitmask(ecx);
```
Comment (tiif) [inline src/shims/unix/linux/epoll.rs:522]: ``get_epoll_ready_events`` will either throw_unsup_format or return an Ok(), Since we use ``unwrap()`` here this function only can be used to check file description that support epoll. Alternatively, we can let this function return an InterpResult and use ? instead of unwrap(). I am not sure which approach is better.

File: src/shims/unix/linux/epoll.rs [hunk 2]
```diff
@@ -525,3 +508,28 @@ fn ready_list_next(
     }
     return None;
 }
+
+/// This helper function checks whether an epoll notification should be triggered for a specific
+/// epoll_interest and, if necessary, triggers the notification. Unlike check_and_update_readiness,
+/// this function sends a notification to only one epoll instance.
+fn check_fd_ready_event(
```
Comment (tiif) [inline src/shims/unix/linux/epoll.rs:515]: I have a hard time coming up with a good naming to differentiate this function from ``check_and_update_readiness``. They essentially do the same thing, but this function only checks one ``epoll_event_interest``, where ``check_and_update_readiness`` checks all ``epoll_event_interest``. As a result, this function can only return notification to one epoll instance where ``check_and_update_readiness`` can return notification to multiple epoll instances.

**Comment by RalfJung** [comment]
> Previously, in epoll_ctl, we used check_and_update_readiness, which is a function that would return notification for all epoll file description. But we shouldn't do that because no notification should be returned if there is no I/O activity between two epoll_wait. 

I am confused by this description. There's only one epoll_wait in the first sentence so the second sentence seems to talk about a different case?

Looking at the testcase I think I understand now, but the PR description doesn't explain the problem. The problem is
- Set up an epoll watching the FD
- Call epoll_wait
- Set up another epoll watching the same FD
- Call epoll_wait on the first epoll. Nothing should be reported!

File: tests/pass-dep/libc/libc-epoll.rs [hunk 3]
```diff
@@ -650,3 +655,46 @@ fn test_ready_list_fetching_logic() {
     let expected_value1 = fd1 as u64;
     check_epoll_wait::<1>(epfd, &[(expected_event1, expected_value1)]);
 }
+
+// epoll has different behaviour in multithreading. We previously check_and_update_readiness the
+// moment a file description is registered in an epoll instance. But this has an unfortunate side
+// effect of returning notification to another epfd that shouldn't receive notification.
+fn test_epoll_ctl_multithread() {
```
Comment (RalfJung) [inline tests/pass-dep/libc/libc-epoll.rs:662]: The problem doesn't seem related to concurrency at all, so why is this test multithreaded?

Comment (tiif) [reply tests/pass-dep/libc/libc-epoll.rs:662]: Right, I discovered this when I am testing it on multi-threaded testcase, but I found a way to write a single-threaded testcase for this.

File: src/shims/unix/linux/epoll.rs (hunk 1 above)
Comment (oli-obk) [reply src/shims/unix/linux/epoll.rs:522]: Oh right, we don't have a test for inserting an FD (that doesn't support epoll) into an epoll FD. Please add such a test. Pretty sure it should hit that unwrap

Comment (tiif) [reply src/shims/unix/linux/epoll.rs]: Yup, it hits the unwrap. ``InterpResult`` is better.

**Review by oli-obk — APPROVED** [review]

**Comment by tiif** [comment]
Rebased and CI is green. 

r? @oli-obk
